### ISSUE TITLE
Crash after deleting part of the animation after motion paste

### ISSUE THREAD

**Issue by dalboris** [dalboris/vpaint#117]
A user reported that VPaint often crashes when deleting part of the animation after motion paste. I should investigate that.

**Comment by dalboris** [comment]
Minimal repro:

1. On frame 1, draw vertex + open edge starting and ending at the same vertex
2. Motion paste to frame 5
3. Soft-delete (Del) vertex at frame 0 (or 5). 

![image](https://user-images.githubusercontent.com/4809739/83549729-0f1a7380-a506-11ea-9ba8-572c41169787.png)

What happens? VPaint crashes. Note that if instead of a soft-delete, we do a hard-delete (Ctrl+Del) on the vertex, then VPaint doesn't crash: it correctly deletes the vertex, the incident open edge, and the incident inbetween vertex and inbetween edge.

What should happen? Good question. Probably something similar as what happens when soft-deleting the vertex in this situation (no animation here, just vertices/edges/faces):

![image](https://user-images.githubusercontent.com/4809739/83550688-86043c00-a507-11ea-8ca8-b1a3ee4bef25.png)

The space-time equivalent for the situation described in the minimal repro would be to merge the vertex and open edge at frame 0 into a closed edge, and to merge the inbetween vertex and inbetween open edge into an inbetween closed edge. That isn't a very trivial operation to do in the general case, and especially it isn't even trivial to verify that such soft-delete is possible in any given situation (the space-time topology much be locally manifold at the vertex).

Note that if instead of the minimal repro, we do the following:
1. Draw open edge, split in the middle (3 vertices, 2 open edges)
2. Motion-paste
3. Soft-delete one of the middle key vertex

What happens: it doesn't crash, but behaves like a hard-delete:

![image](https://user-images.githubusercontent.com/4809739/83551659-ec3d8e80-a508-11ea-91f9-0f6b29b68677.png)

Doing a hard-delete is probably the "simplest/safest" thing to do, when there are incident inbetween cells. Although it might be nice to do something smarter like the soft-delete in the non-animated example.

**Comment by dalboris** [comment]
Alright, this issue, and possibly other related issues, are now partially fixed. More precisely, deleting cells should now never crash: it might delete more than we expect (example: all incident inbetween cells), or less than we expect (example: soft-deleting inbetween edges does nothing), but at least it should not crash VPaint.

The bug was that the uncut operator, which was initially implemented for fixed illustrations, was not properly generalized to support animation. Such generalization was properly implemented for the cut and the glue operators, only partially implemented for unglue (see below), and not at all implemented for uncut (it seems to have been an oversight).

Generalizing unglue and uncut to animation is complicated (more complicated than cut and glue), and I haven't yet taken the time to properly study and implement these generalizations. In the case of unglue, while a proper generalization is not implemented, there was at least already a failsafe: all incident inbetween cells are deleted before performing the operation, which might be unexpected, but at least doesn't cause a crash and does what's expected in the current frame.

Such failsafe was not even implemented for uncut, but I just added it. The behavior is now that the uncut operator returns "impossible to uncut" whenever there are inbetween cells incident to the selected cell. Consequently, this means that a soft-delete of any key cell with incident inbetween cells always become a hard-delete. In the future, we should instead do something smarter, and actually detect cases where both the key cell and its incident inbetween cells can be uncut, and change the topology accordingly.

Note that before the failsafe was implemented, it was just "luck" that in many cases, uncutting/soft-deleting didn't crash. Inbetween cells where often indirectly deleted as a result of deleting the old key cells to merge. But in the specific case of uncutting at a key vertex with an incident looping key open edge, the existing incident key edge is not deleted: it is "converted" to a closed edge. As a result, the incident inbetween edge was not deleted neither made aware of this conversion, and later caused a crash when attempting to dereference the end-vertices of the now-closed key edge.
